### PR TITLE
Apply synchronization over network IR to protect busy-resources list

### PR DESCRIPTION
**Pull request by d4vidi**
## Description

- This is a follow-up on this contribution: #3608

In this pull request, I have reimplemented #3608 in a more robust way.

Co-authored-by: @zhouzh1 <zzhouzhii@gmail.com>

---

> _For features/enhancements:_
 - [ ] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [ ] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

### CODE REVIEW

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/NetworkIdlingResource.java [hunk 1]
```diff
@@ -116,10 +87,42 @@ public void doFrame(long frameTimeNanos) {
         isIdleNow();
     }
 
+    @Override
+    protected synchronized boolean checkIdle() {
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/NetworkIdlingResource.java:91]: Also `synchronized`

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/NetworkIdlingResource.java [hunk 2]
```diff
@@ -116,10 +87,42 @@ public void doFrame(long frameTimeNanos) {
...
+
+        if (!busyResources.isEmpty()) {
+            Log.i(LOG_TAG, "Network is busy, with " + busyResources.size() + " in-flight calls");
+            Choreographer.getInstance().postFrameCallback(this);
+            return false;
+        }
+
+        notifyIdle();
+        return true;
+    }
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/NetworkIdlingResource.java:111]: We should rewrite this in Kotlin. This is terrible.

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/NetworkIdlingResource.java [hunk 3]
```diff
@@ -69,42 +69,13 @@ public String getName() {
 
     @NotNull
     @Override
-    public IdlingResourceDescription getDescription() {
+    public synchronized IdlingResourceDescription getDescription() {
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/NetworkIdlingResource.java:72]: `synchronized` 1

**Review by asafkorem — COMMENTED** [review]
Good stuff :shipit:

**Pull request merged**